### ISSUE TITLE
Subscribe as state data provider when state is uninitialized

### ISSUE THREAD

**Issue by gkatsikas** [sysrepo/sysrepo#2154]
Hi,

I have a yang model which organizes device data in a list.
Each device has a config and a state container, but in order to access these 2 containers, one should first create a list item for this device (i.e, /device-list[id="0"]) and then access the state (i.e., /device-list[id="0"]/state/variable).

Now let's got to the problem.
When I start my plugin, I have to subscribe as state data provider for (all) the devices' state data, but initially there is no device in the list.
In the oven example, that was easily done by subscribing to "/oven:oven-state" (there is only one oven).
In my case, this path appears in the tree, when one calls a custom RPC to add a new device.

Is this an accepted approach or is there a fundamental design problem in my yang model?
In other words, if someone wants to manage multiple oven devices, how would you do that? By providing a yang model which lists multiple oven devices or by deploying multiple oven plugins across multiple devices (one device per oven)? In the latter case, there are multiple sysrepos and this is something I would like to avoid.

Thanks in advance,
Georgios

**Comment by michalvasko** [comment]
Hi Georgios,

> When I start my plugin, I have to subscribe as state data provider for (all) the devices' state data, but initially there is no device in the list.

You can subscribe to all list instances in the path, address it as if you were addressing a container (do not use any predicates).

> In my case, this path appears in the tree, when one calls a custom RPC to add a new device.

Unless there is a specific reason why the configuration cannot be set directly but only using a custom RPC, you should avoid it because it can be confusing if the data are `config` `true` but cannot be configured.

> Is this an accepted approach or is there a fundamental design problem in my yang model?
In other words, if someone wants to manage multiple oven devices, how would you do that? By providing a yang model which lists multiple oven devices or by deploying multiple oven plugins across multiple devices (one device per oven)? In the latter case, there are multiple sysrepos and this is something I would like to avoid.

The former solution in case all the devices are the same and their management code should be the same (which is expected if they are a different instances of some configuration subtree).

**Comment by gkatsikas** [comment]
Hi Michal,

Thanks a lot for the feedback! Indeed the indexing of the container works as you suggested.
